### ISSUE TITLE
Text-Layer has wrong position in some fonts

### ISSUE THREAD

**Issue by Fearodin** [mozilla/pdf.js#15114]
Attach (recommended) or Link to PDF file here:
[1000-Seiten.pdf](https://github.com/mozilla/pdf.js/files/9009977/1000-Seiten.pdf)
h

Configuration:
- Web browser and its version: Firefox 102.0  (64bit)
- Operating system and its version: Windows 10 21H2 19044.1706
- PDF.js version: as in latest Firefox
- Is a browser extension: no

Steps to reproduce the problem:
1. Open the attached PDF in Firefox or any pdf.js versions (tried some)
2. Move to page 900
3. Select the text "Gemüsebrühe" in the lower area of the page
![image](https://user-images.githubusercontent.com/12547599/176425804-e840e37a-b836-4b4e-a41c-91f2edcbc5e4.png)

What is the expected behavior? (add screenshot)
The text-layer should lay exactly over the visible text, like it is in chrome or adobe:  
![image](https://user-images.githubusercontent.com/12547599/176425896-a7c19eaf-9744-4124-ae84-95a626056b08.png)

What went wrong? (add screenshot)
The text-layer is way beside the visible text. I think there is a problem resolving the font, which is used in the pdf.
In this screenshot I moved the textlayer of the target section a bit to the bottom to get the problem visible:
![image](https://user-images.githubusercontent.com/12547599/176426299-e4e6209f-3b42-4581-b614-5d4ce4be1fc2.png)

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):
Just try the scenario with the attached PDF in any pdf-viewer you like (Chrome, Adobe, Foxit). The problem only occurs in Firefox.

**Comment by orbiteleven** [comment]
@Fearodin did you ever find a solution for this? Seeing the same 2 years later.

**Comment by Fearodin** [comment]
> @Fearodin did you ever find a solution for this? Seeing the same 2 years later.

@orbiteleven nope. Still there in the latest firefox-versions and pdf.js-versions. Seems to be an issue of how the embedded fonts are resolved.

**Comment by BingCoke** [comment]
> > @Fearodin did you ever find a solution for this? Seeing the same 2 years later.
> 
> @orbiteleven nope. Still there in the latest firefox-versions and pdf.js-versions. Seems to be an issue of how the embedded fonts are resolved. 

i have read textlayer code,has no problem.So think it may be not in canvas? So I used the same logic to render in canvas and got the same result.

<img width="765" alt="截屏2024-07-08 19 57 03" src="https://github.com/mozilla/pdf.js/assets/81607010/8c390354-d34a-49c8-9d4a-3fdf40d09204">
But I realized that the font seemed wrong, and the rendered letter a was completely different.
pdfjs will extract textContext which contains textItem. textItem has fontName property,and textContent has styles which is a map ,key is fontName,value is fontFamily. I extract pdf fonts use pdfcpu command get  TimesNewRoman.ttf. but pdfjs extract  textContext's font is serif!
<img width="559" alt="截屏2024-07-08 20 01 37" src="https://github.com/mozilla/pdf.js/assets/81607010/a3d587e1-7b98-40da-b638-ff9a88425a2e">
after i change all of fontFamily to "timesNewRoman" everything is ok.
<img width="470" alt="截屏2024-07-08 20 08 22" src="https://github.com/mozilla/pdf.js/assets/81607010/3bb47a3f-8b8e-4b56-80c9-0771d661cbc5">

Because there is an operation to obtain Ascent during rendering to ensure that the font position and width are adjusted. Because the obtained font data is wrong, it will cause the rendered font to have slight expansion and contraction changes.
I don't have much energy to solve this problem yet. If the problem has not been solved when I have time, I will try to solve it.
